### PR TITLE
SALTO-5932: create field context options elements

### PR DESCRIPTION
**Pull request by SeggevHaimovich**
We created stand alone elements from the options in the contexts of fields. We do so, to enable modification of options instead of deletion and recreation.
This will happen if the flag `splitFieldContext` in the config will be enabled.

Deploying options will remain under deploying context. It means that changes in options will result in deploying the context.

---

_Additional context for reviewer_
We did something similar with `fieldConfigurationItems` in this PR: https://github.com/salto-io/salto/pull/3006

---
_Release Notes_: 
Jira Adapter:
* changed the structure of field context to enable modification of options, options will have separated elements

---
_User Notifications_: 
_field contexts will now have a list of references to their options, options will now be in standalone elements_

### CODE REVIEW

**Comment by SeggevHaimovich** [comment]
I still haven't finished the tests and the CV to check for duplicate name of options within the same context

File: packages/jira-adapter/src/filters/fields/context_deployment_filter.ts [hunk 1]
```diff
@@ -132,15 +233,18 @@ const filter: FilterCreator = ({ client, config, paginator, elementsSource }) =>
             return
           }
         }
-        await deployContextChange(change, client, config.apiDefinitions, paginator, elementsSource)
+        const changeToDeploy = getContextChangeWithFixedOptions(change, optionFullNameToChangeDict)
```
Comment (IdoZakk) [inline packages/jira-adapter/src/filters/fields/context_deployment_filter.ts:236]: We usually prefer to make changes in a dedicated preDeploy filter (and reverse them in post deploy). It makes the deployment code simpler. 
Another option (which will make much more sense) is to change the function `setContextOptions` in `context_options.ts`. We currently look at before and after to deduce changes, and we already have this information

File: packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  let elements: InstanceElement[]
+  let contextInstance: InstanceElement
+  let changes: ReadonlyArray<Change<ChangeDataType>>
+  let optionType: ObjectType
+  let optionInstance1: InstanceElement
+  let optionInstance2: InstanceElement
+
+  beforeEach(() => {
+    jest.clearAllMocks()
+    contextType = new ObjectType({ elemID: new ElemID(JIRA, FIELD_CONTEXT_TYPE_NAME) })
```
Comment (IdoZakk) [inline packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts:44]: Nit: you can use `createEmptyType`

**Review by IdoZakk — COMMENTED** [review]
Nice Work 🥳 🎈 🪓 
Some general points:
* You did not handle the default option in the context. When fetching my env I can see I get a warning about it. As a practice you should use your code to fetch ori-salto-test and make sure you did not break anything there. It is annoying has there are several fetch errors there that you should ingore
* Reading this PR was a bit rough. Next time please try to create a different PR for fetch and for deploy
* Please test a context with more than 10K options and make sure it works fine
* You did not handle cascading options
* There is a problem in an edge case we might solve easily. Not sure it is worth working on if solution is not trivial. The problem is when you rename option 'a' to 'b', and then add another option 'a'. It creates a merge problem, which is expected, but the references in the context are not missing references. I think that you can check for it in the split filter and issue a warning and add missing references links
* I wrote in the comments, I think we can simplify the deploy. Also the deploy code went some changes (clean up), so you might want to rebase

File: packages/jira-adapter/src/filters/fields/field_context_option_split.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+export const CONTEXT_NAME_FIELD = 'contextName'
+
+const getOptionInstancesFromContext = async (
+  context: InstanceElement,
+  fieldContextOptionType: ObjectType,
+  getElemIdFunc: ElemIdGetter | undefined,
+): Promise<InstanceElement[]> =>
+  Promise.all(
+    getOptionsFromContext(context).map(async (option: Values) => {
+      option[CONTEXT_NAME_FIELD] = context.elemID.name
```
[the remark this answers is not in the document]
Comment (SeggevHaimovich) [reply packages/jira-adapter/src/filters/fields/field_context_option_split.ts:51]: I am going to update it to `parentName`
I use it to create the elemID.
In the `api_config.ts` file I defined the `idFields` as ['parentName', 'value'].

File: packages/jira-adapter/src/filters/fields/field_context_option_split.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      const optionInstance = await toBasicInstance({
+        entry: option,
+        type: fieldContextOptionType,
+        transformationConfigByType: getTransformationConfigByType(DEFAULT_API_DEFINITIONS.types),
+        transformationDefaultConfig: DEFAULT_API_DEFINITIONS.typeDefaults.transformation,
+        defaultName: naclCase(`${context.elemID.name}_${option.value}`),
+        getElemIdFunc,
+        parent: context,
+      })
+      optionInstance.path = context.path && [
```
[the remark this answers is not in the document]
Comment (SeggevHaimovich) [reply packages/jira-adapter/src/filters/fields/field_context_option_split.ts:61]: After the change I do now, the path is from the context and not necessarily from the parent (cascadingOption's parent is an option but the path is still under the context's folder)

File: packages/jira-adapter/src/filters/fields/context_deployment_filter.ts [hunk 5]
```diff
@@ -99,6 +117,80 @@ const deployJsmContextField = async (
...
+  optionRef: ReferenceExpression,
+  index: number,
+  optionFullNameToChangeRecord: Record<string, Change<InstanceElement>>,
+  beforeOrAfter: BeforeOrAfter,
+): [string, Values] | undefined => {
+  // const option = Object.prototype.hasOwnProperty.call(optionFullNameToChangeRecord, optionRef.elemID.getFullName())
+  //   ? _.get(optionFullNameToChangeRecord[optionRef.elemID.getFullName()].data, beforeOrAfter)
+  //   : optionRef.value
+  // const optionValue = _.omit(option.value, CONTEXT_NAME_FIELD)
+  // return [optionValue.value, { ...optionValue, position: index + 1 }]
```
[the remark this answers is not in the document]
Comment (SeggevHaimovich) [reply packages/jira-adapter/src/filters/fields/context_deployment_filter.ts:130]: actually I left it because it does the same without replicating the lines:
`const optionValue = _.omit(option.value, CONTEXT_NAME_FIELD)`
`return [optionValue.value, { ...optionValue, position: index + 1 }]`
but it felt less readable

WDYT?

**Review by IdoZakk — COMMENTED** [review]
Better to close this PR and focus on the other one
